File: iosApp/Shared/Screens/MyProfile/Availability/RequestConsultationView.swift
//
//  RequestConsultationView.swift
//  Well
//
//  Created by Phil on 28.10.2021.
//  Copyright © 2021 Well. All rights reserved.
//

import SwiftUI
import SharedMobile

private typealias Feature = RequestConsultationFeature

struct RequestConsultationOverlay: View {
    let state: RequestConsultationFeature.State
    
    var body: some View {
        if isVisible {
            InactiveOverlay {
                ZStack {
                    switch state.status {
                    case is Feature.StateStatusProcessing:
                        ProgressView()
                            .progressViewStyle(CircularProgressViewStyle(tint: .companion.Green))

                    case is Feature.StateStatusBooked:
                        Image(systemName: "checkmark")
                            .font(.system(size: 60, weight: .black))
                            .foregroundColorKMM(.companion.Green)
                            .padding().background(RoundedRectangle(cornerRadius: 16).foregroundColorKMM(.companion.White))

                    case let failed as Feature.StateStatusBookingFailed:
                        VStack {
                            Text(Feature.Strings.shared.bookingFailed)
                            Text(failed.reason)
                        }.padding().background(RoundedRectangle(cornerRadius: 16).foregroundColorKMM(.companion.White))

                    case is Feature.StateStatusLoading, is Feature.StateStatusLoaded:
                        EmptyView()
                    default:
                        EmptyView()
                    }
                }.padding()
                    .background(
                        RoundedRectangle(cornerRadius: 16)
                            .foregroundColor(.white)
                    )
            }.animation(.default, value: state.status)
        }
    }

    private var isEmptyContent: Bool {
        switch state.status {
        case
            is Feature.StateStatusProcessing,
            is Feature.StateStatusLoading,
            is Feature.StateStatusLoaded:

            return true

        default:
            return false
        }
    }

    private var isVisible: Bool {
        switch state.status {
        case
            is Feature.StateStatusProcessing,
            is Feature.StateStatusBooked,
            is Feature.StateStatusBookingFailed:

            return true

        default:
            return false
        }
    }
}

struct RequestConsultationView: View {
    let state: RequestConsultationFeature.State
    let listener: (RequestConsultationFeature.Msg) -> Void
    
    var body: some View {
        VStack {
            content
        }
    }
    
    @ViewBuilder
    var content: some View {
        Spacer().frame(height: 16)
        Text(Feature.Strings.shared.title)
            .textStyle(.subtitle2)
        if state.status is Feature.StateStatusLoading {
            ProgressView()
                .progressViewStyle(CircularProgressViewStyle(tint: .companion.Green))
        } else if state.availabilitiesByDay.isEmpty {
            Text(Feature.Strings.shared.hasNoConsultations)
        } else {
            VStack(spacing: 52) {
                Availabilities(
                    availabilities: state.availabilitiesByDay,
                    onBookNow: {
                        listener(Feature.MsgBook(availability: $0))
                    }
                )
            }
        }
    }
}

private struct Availabilities: View {
    let availabilities: [KotlinPair<LocalDate, NSArray>]
    let onBookNow: (BookingAvailability) -> Void
    
    @State
    private var selectedDayIndex = 0
    
    @State
    private var selectedTimeIndex = -1
    
    var body: some View {
        VStack(spacing: 28) {
            ScrollViewReader { scrollView in
                BookRow(
                    items: availabilities.map { $0.first! },
                    id: \.self,
                    selectedIndex: $selectedDayIndex,
                    itemText: { $0.localizedDayAndShortMonth(separator: "\n") },
                    textStyle: .body2,
                    aspectRatio: 1,
                    parentPart: 0.2
                ).onChange(of: selectedDayIndex) { _ in
                    selectedTimeIndex = -1
                }.onChange(of: selectedTimeIndex) { _ in
                    withAnimation {
                        scrollView.scrollTo(selectedDayIndex)
                    }
                }
            }
            Divider().padding(.horizontal)
            BookRow(
                items: availabilities[selectedDayIndex].second as! [BookingAvailability],
                id: \.startTime,
                selectedIndex: $selectedTimeIndex,
                itemText: { $0.startTime.toString() },
                textStyle: .body1,
                aspectRatio: 2,
                parentPart: 0.33
            )
        }
        Button(
            item: selectedTimeIndex,
            disabled: { $0 < 0 },
            action: { unwrappedSelectedTimeIndex in
                onBookNow(availabilities[selectedDayIndex].second![unwrappedSelectedTimeIndex] as! BookingAvailability)
            }
        ) {
            Text(Feature.Strings.shared.bookNow)
        }.buttonStyle(ActionButtonStyle(style: .onWhite)).padding(.horizontal)
        // to apply padding by verticalArrangement
        Spacer().height(0)
    }
}

private struct BookRow<T, ID: Hashable>: View {
    let items: [T]
    let id: KeyPath<T, ID>
    @Binding
    var selectedIndex: Int

    let itemText: (T) -> String
    let textStyle: TextStyle
    let aspectRatio: Double
    let parentPart: CGFloat
    
    @State
    var height: CGFloat?
    
    var body: some View {
        let width: CGFloat = (UIScreen.main.bounds.width * parentPart).rounded()
        let itemSize = CGSize(width: width, height: (width / aspectRatio).rounded())
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 8) {
                ForEachIndexed(items) { i, item in
                    let text = itemText(item)
                    SelectableButton(
                        selected: i == selectedIndex,
                        onClick: {
                            withAnimation {
                                selectedIndex = i
                            }
                        },
                        content: Text(text)
                            .frame(size: itemSize)
                            .textStyle(textStyle)
                            .multilineTextAlignment(.center)
                    )
                    .id(text)
                }
            }.padding(.horizontal)
        }
    }
}

File: modules/androidUi/src/main/kotlin/com/well/modules/androidUi/composableScreens/myProfile/availability/RequestConsultationView.kt
package com.well.modules.androidUi.composableScreens.myProfile.availability

import com.well.modules.androidUi.components.ActionButton
import com.well.modules.androidUi.components.BottomSheetDialog
import com.well.modules.androidUi.components.InactiveOverlay
import com.well.modules.androidUi.components.SelectableButton
import com.well.modules.androidUi.ext.toColor
import com.well.modules.features.myProfile.myProfileFeature.availabilitiesCalendar.RequestConsultationFeature.Msg
import com.well.modules.features.myProfile.myProfileFeature.availabilitiesCalendar.RequestConsultationFeature.State
import com.well.modules.models.BookingAvailability
import com.well.modules.models.Color
import com.well.modules.models.date.dateTime.localizedDayAndShortMonth
import com.well.modules.features.myProfile.myProfileFeature.availabilitiesCalendar.RequestConsultationFeature as Feature
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyListState
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material.CircularProgressIndicator
import androidx.compose.material.Divider
import androidx.compose.material.Icon
import androidx.compose.material.MaterialTheme
import androidx.compose.material.ProvideTextStyle
import androidx.compose.material.Text
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.launch
import kotlinx.datetime.LocalDate

@Composable
fun RequestConsultationBottomSheet(
    state: State,
    listener: (Msg) -> Unit,
) {
    BottomSheetDialog(onDismissRequest = { listener(Msg.Close) }) {
        Box {
            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
                modifier = Modifier
                    .fillMaxWidth()
            ) {
                RequestConsultationView(state, listener)
            }
            Overlay(state.status)
        }
    }
}

@Composable
private fun BoxScope.Overlay(status: State.Status) {
    when (status) {
        State.Status.Loading,
        State.Status.Loaded,
        -> Unit
        State.Status.Processing -> {
            InactiveOverlay(showActivityIndicator = true)
        }
        State.Status.Booked -> {
            InactiveOverlay() {
                Icon(
                    Icons.Default.Check,
                    contentDescription = null,
                    modifier = Modifier
                        .fillMaxWidth(0.5f)
                        .aspectRatio(1f),
                    tint = Color.Green.toColor()
                )
            }
        }
        is State.Status.BookingFailed -> InactiveOverlay() {
            Column(horizontalAlignment = Alignment.CenterHorizontally) {
                Text(Feature.Strings.bookingFailed, style = MaterialTheme.typography.body1)
                Text(status.reason, textAlign = TextAlign.Center)
            }
        }
    }
}

@Composable
private fun RequestConsultationView(
    state: State,
    listener: (Msg) -> Unit,
) {
    Spacer(Modifier.height(16.dp))
    Text(
        Feature.Strings.title,
        style = MaterialTheme.typography.subtitle2,
    )
    Spacer(Modifier.height(40.dp))
    when {
        state.status == State.Status.Loading -> {
            CircularProgressIndicator(
                modifier = Modifier
                    .fillMaxWidth(0.1f)
                    .aspectRatio(1f)
            )
            Spacer(Modifier.height(40.dp))
        }
        state.availabilitiesByDay.isEmpty() -> {
            Text(Feature.Strings.hasNoConsultations, style = MaterialTheme.typography.body1)
            Spacer(Modifier.height(40.dp))
        }
        else -> {
            Column(
                verticalArrangement = Arrangement.spacedBy(52.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                Availabilities(state.availabilitiesByDay) {
                    listener(Msg.Book(it))
                }
            }
        }
    }
}

@Composable
private fun Availabilities(
    availabilities: List<Pair<LocalDate, List<BookingAvailability>>>,
    onBookNow: (BookingAvailability) -> Unit,
) {
    val (selectedDayIndex, setSelectedDayIndex) = rememberSaveable(availabilities) {
        mutableStateOf(0)
    }
    val (selectedTimeIndex, setSelectedTimeIndex) = rememberSaveable(selectedDayIndex) {
        mutableStateOf<Int?>(null)
    }
    val scope = rememberCoroutineScope()
    val topRowState = rememberLazyListState()
    Column(
        verticalArrangement = Arrangement.spacedBy(28.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        ProvideTextStyle(MaterialTheme.typography.body2) {
            BookRow(
                state = topRowState,
                items = remember(availabilities) { availabilities.map { it.first } },
                selectedIndex = selectedDayIndex,
                setSelectedIndex = setSelectedDayIndex,
                itemText = { it.localizedDayAndShortMonth(separator = "\n") },
                aspectRatio = 1f,
                fillParentMaxWidthPart = 0.2f
            )
        }
        Divider(Modifier.padding(horizontal = 16.dp))

        ProvideTextStyle(MaterialTheme.typography.body1) {
            BookRow(
                items = remember(selectedDayIndex) { availabilities[selectedDayIndex].second },
                selectedIndex = selectedTimeIndex,
                setSelectedIndex = { selectedIndex ->
                    setSelectedTimeIndex(selectedIndex)
                    if (topRowState.layoutInfo.visibleItemsInfo.none { it.index == selectedDayIndex }) {
                        scope.launch {
                            try {
                                topRowState.animateScrollToItem(selectedDayIndex)
                            } catch (_: Throwable) {
                            }
                        }
                    }
                },
                itemText = { it.startTime.toString() },
                aspectRatio = 2f,
                fillParentMaxWidthPart = 0.33f
            )
        }
    }
    ActionButton(
        state = selectedTimeIndex,
        onClick = { unwrappedSelectedTimeIndex ->
            scope.launch {
                if (topRowState.layoutInfo.visibleItemsInfo.any { it.index == 0 }) {
                    topRowState.animateScrollToItem(availabilities.count() - 1)
                } else {
                    topRowState.scrollToItem(0)
                }
            }
            onBookNow(availabilities[selectedDayIndex].second[unwrappedSelectedTimeIndex])
        },
        modifier = Modifier.padding(horizontal = 16.dp)
    ) {
        Text(Feature.Strings.bookNow)
    }
    // to apply padding by verticalArrangement
    Spacer(Modifier)
}

@Composable
private fun <T> BookRow(
    state: LazyListState = rememberLazyListState(),
    items: List<T>,
    selectedIndex: Int?,
    setSelectedIndex: (Int) -> Unit,
    itemText: (T) -> String,
    aspectRatio: Float,
    fillParentMaxWidthPart: Float,
) {
    LazyRow(
        state = state,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        contentPadding = PaddingValues(horizontal = 16.dp),
        modifier = Modifier.fillMaxWidth()
    ) {
        itemsIndexed(items) { i, item ->
            SelectableButton(
                selected = i == selectedIndex,
                onClick = {
                    setSelectedIndex(i)
                },
                modifier = Modifier
                    .fillParentMaxWidth(fillParentMaxWidthPart)
                    .aspectRatio(aspectRatio)
            ) {
                Text(
                    itemText(item),
                    textAlign = TextAlign.Center,
                )
            }
        }
    }
}